test/Microsoft.ML.Functional.Tests/DataTransformation.cs

```diff
@@ -0,0 +1,243 @@
+            foreach (var row in transformedRows)
+            {
+                var randomNumber = GetRandomNumber(row.Label + row.PetalLength + row.PetalWidth + row.SepalLength + row.SepalWidth);
+                Assert.Equal(randomNumber, row.Float1);
+            }
+        }
+
+        /// <summary>
+        /// Extensibility: Add multiple new columns.
+        /// </summary>
+        [Fact]
+        void ExtensibilityAddingTwoColumns()
+        {
+            // Concurrency must be 1 to assure that the mapping is done sequentially.
+            var mlContext = new MLContext(seed: 1, conc: 1);
+
+            // Load the Iris dataset
+            var data = mlContext.Data.LoadFromTextFile<Iris>(
+                GetDataPath(TestDatasets.iris.trainFilename),
+                hasHeader: TestDatasets.iris.fileHasHeader,
+                separatorChar: TestDatasets.iris.fileSeparator);
+
+            // Subsample it down to the first 10 rows.
+            int numSamples = 10;
+            data = mlContext.Data.TakeRows(data, numSamples);
+
+            // Create a function that generates a column.
+            Action<Iris, IrisWithTwoExtraColumns> generateGroupId = (input, output) =>
+            {
+                output.Label = input.Label;
+                output.Float1 = GetRandomNumber(1 + input.Label + input.PetalLength + input.PetalWidth + input.SepalLength + input.SepalWidth);
+                output.Float2 = GetRandomNumber(2 + input.Label + input.PetalLength + input.PetalWidth + input.SepalLength + input.SepalWidth);
+                output.PetalLength = input.PetalLength;
+                output.PetalWidth = input.PetalWidth;
+                output.SepalLength = input.SepalLength;
+                output.SepalWidth = input.SepalWidth;
+            };
+
+            // Create a pipeline to execute the custom function.
+            var pipeline = mlContext.Transforms.CustomMapping(generateGroupId, null);
+
+            // Transform the data.
+            var transformedData = pipeline.Fit(data).Transform(data);
+
+            // Verify that the column has the correct data.
+            var transformedRows = mlContext.Data.CreateEnumerable<IrisWithTwoExtraColumns>(transformedData, reuseRowObject: true);
+            foreach (var row in transformedRows)
+            {
+                var randomNumber1 = GetRandomNumber(1 + row.Label + row.PetalLength + row.PetalWidth + row.SepalLength + row.SepalWidth);
+                var randomNumber2 = GetRandomNumber(2 + row.Label + row.PetalLength + row.PetalWidth + row.SepalLength + row.SepalWidth);
+                Assert.Equal(randomNumber1, row.Float1);
+                Assert.Equal(randomNumber2, row.Float2);
+            }
+        }
+
+        /// <summary>
+        /// Extensibility: Add a new column that is a function of other columns.
+        /// </summary>
+        [Fact]
+        void ExtensibilityAddAColumnAsAFunctionOfMultipleColumns()
```

Comment:
>ExtensibilityAddAColumnAsAFunctionOfMultipleColumns [](start = 13, length = 51)

to me this test seems identical to the ExtensibilityAddingOneColumn. The function is different, but does that matter to the CustomMapping?

Reply:
I agree, maybe I would only keep one of the two functions.

---
In reply to: [261830627](https://github.com/dotnet/machinelearning/pull/2803#discussion_r261830627) [](ancestors = 261830627)